### PR TITLE
DX: check that tests are agnostic of PHP version

### PR DESCRIPTION
**Pull request by kubawerlos**
Replaces https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7575

### CODE REVIEW

File: tests/AutoReview/ProjectCodeTest.php [hunk 1]
```diff
@@ -763,12 +818,15 @@ public function testConstantsAreInUpperCase(string $className): void
 
     private function createTokensForClass(string $className): Tokens
     {
-        $file = $className;
-        $file = preg_replace('#^PhpCsFixer\\\Tests\\\#', 'tests\\', $file);
-        $file = preg_replace('#^PhpCsFixer\\\#', 'src\\', $file);
-        $file = str_replace('\\', \DIRECTORY_SEPARATOR, $file).'.php';
+        $filename = (new \ReflectionClass($className))->getFileName();
```
Comment (keradus) [inline tests/AutoReview/ProjectCodeTest.php:821]: back to few days ago we were using both approaches, filename to filepath manually and with reflection class. tbh, I would like to avoid reflection usage here. It is some kind of "let's do magic under the hood" when we are almost abusing the language, for me. filesystem approach is "non-magic" - we do not need to rely on some internals to tell us where is the file, file is ours, we have strict structure on filesystem, we know where to access any file for any class.

File: tests/PregTest.php [hunk 2]
```diff
@@ -86,7 +84,7 @@ public function testPatternValidation(string $pattern, ?bool $expected = null, ?
 
             if (null !== $expectedMessage) {
                 ++$i;
-                $this->expectExceptionMessage($expectedMessage);
+                $this->expectExceptionMessageMatches($expectedMessage);
```
Comment (keradus) [inline tests/PregTest.php:87]: why the change?

File: tests/Tokenizer/Transformer/TypeIntersectionTransformerTest.php [hunk 3]
```diff
@@ -63,7 +63,7 @@ function foo(){}
             ',
         ];
 
-        if (\PHP_VERSION_ID >= 8_01_00) {
+        if (\defined('T_AMPERSAND_FOLLOWED_BY_VAR_OR_VARARG')) { // @TODO: drop condition when PHP 8.1+ is required
```
Comment (keradus) [inline tests/Tokenizer/Transformer/TypeIntersectionTransformerTest.php:66]: for me, it is as unfortunate as checking PHP version directly , with very same reasoning. let's have dedicated provider.

File: tests/AutoReview/ProjectCodeTest.php [hunk 4]
```diff
@@ -348,6 +348,61 @@ public function testThatTestClassCoversAreCorrect(string $testClassName): void
         }
     }
 
+    /**
+     * @dataProvider provideTestClassCases
+     */
+    public function testThereIsNoPhPVersionUsedDirectly(string $className): void
```
Comment (keradus) [inline tests/AutoReview/ProjectCodeTest.php:354]: `PhP`

File: tests/PregTest.php (hunk 2 above)
Comment (kubawerlos) [reply tests/PregTest.php:87]: Removed line:

> $nullByteMessage = \PHP_VERSION_ID >= 8_02_00 ? 'NUL is not a valid modifier' : 'Null byte in regex';

We don't have to expect the message based on the PHP version, we can simply check if the message matches any of these two.

File: tests/AutoReview/ProjectCodeTest.php [hunk 5]
```diff
@@ -763,12 +818,15 @@ public function testConstantsAreInUpperCase(string $className): void
 
     private function createTokensForClass(string $className): Tokens
     {
-        $file = $className;
-        $file = preg_replace('#^PhpCsFixer\\\Tests\\\#', 'tests\\', $file);
-        $file = preg_replace('#^PhpCsFixer\\\#', 'src\\', $file);
-        $file = str_replace('\\', \DIRECTORY_SEPARATOR, $file).'.php';
+        $filename = (new \ReflectionClass($className))->getFileName();
+
+        static $fileTokens = [];
```
Comment (keradus) [inline tests/AutoReview/ProjectCodeTest.php:823]: I like the cache. Let's clear it in `tearDownAfterClass` ?

File: tests/Tokenizer/Analyzer/Analysis/TypeAnalysisTest.php [hunk 6]
```diff
@@ -123,75 +126,120 @@ public static function provideIsNullableCases(): iterable
...
+        yield [false, 'string|Null2'];
 
-            yield [false, 'Null2| int'];
+        yield [false, 'string |Null2'];
 
-            yield [false, 'string | Null2 | int'];
-        }
+        yield [false, 'Null2| int'];
 
-        if (\PHP_VERSION_ID >= 8_01_00) {
```
Comment (keradus) [inline tests/Tokenizer/Analyzer/Analysis/TypeAnalysisTest.php:175]: good to have this moved to standalone PR. let's take care of that standalone PR first

File: tests/Tokenizer/Transformer/TypeIntersectionTransformerTest.php (hunk 3 above)
Comment (kubawerlos) [reply tests/Tokenizer/Transformer/TypeIntersectionTransformerTest.php:66]: any idea how to do it right?

File: tests/AutoReview/ProjectCodeTest.php [hunk 7]
```diff
@@ -348,6 +348,61 @@ public function testThatTestClassCoversAreCorrect(string $testClassName): void
...
+        );
+
+        $strings = array_map(
+            static fn (Token $token): string => $token->getContent(),
+            $stringTokens
+        );
+
+        $strings = array_unique($strings);
+        $message = sprintf('%s must not used directly.', $className);
+        self::assertNotContains('phpversion', $strings, $message);
```
Comment (keradus) [inline tests/AutoReview/ProjectCodeTest.php:395]: i'm not convinced to prevent usage in _whole_ test file
ref https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7575#issuecomment-1858894164

File: tests/Tokenizer/Transformer/TypeIntersectionTransformerTest.php (hunk 3 above)
Comment (kubawerlos) [reply tests/Tokenizer/Transformer/TypeIntersectionTransformerTest.php:66]: Maybe the next iteration of refactoring could disallow `defined` in tests as well? Not much we have:
```
kub@:~/code/friendsofphp/PHP-CS-Fixer(master)$ git grep 'defined(' ./tests
tests/AutoReview/ProjectCodeTest.php:            if (\defined('T_ATTRIBUTE') && $tokens[$index]->isGivenKind(T_ATTRIBUTE)) {
tests/Fixer/Alias/NoAliasFunctionsFixerTest.php:            if (defined("'.$alias.'") || $'.$alias.' instanceof '.$alias.') {
tests/Fixer/Alias/RandomApiMigrationFixerTest.php:        if (!defined("srand") || $srand instanceof srand) {
tests/Fixer/Basic/PsrAutoloadingFixerTest.php:            if (\defined($tokenType)) {
tests/Tokenizer/Analyzer/ControlCaseStructuresAnalyzerTest.php:        if (\defined('T_ENUM')) { // @TODO: drop condition when PHP 8.1+ is required - sadly PHPUnit still calls the provider even if requires condition is not matched
tests/Tokenizer/Analyzer/ControlCaseStructuresAnalyzerTest.php:        if (\defined('T_MATCH')) { // @TODO: drop condition when PHP 8.0+ is required - sadly PHPUnit still calls the provider even if requires condition is not matched
tests/Tokenizer/CTTest.php:        self::assertFalse(\defined($name), 'The CT name must not use native T_* name.');
tests/Tokenizer/TokenTest.php:        if (\defined('T_ATTRIBUTE')) {
tests/Tokenizer/TokenTest.php:        if (\defined('T_NULLSAFE_OBJECT_OPERATOR')) {
tests/Tokenizer/TokenTest.php:        if (\defined('T_AMPERSAND_FOLLOWED_BY_VAR_OR_VARARG')) { // @TODO: drop condition when PHP 8.1+ is required
tests/Tokenizer/TokenTest.php:        if (\defined('T_ENUM')) {
tests/Tokenizer/Transformer/NameQualifiedTransformerTest.php:        if (\defined('T_NAME_RELATIVE')) { // @TODO: drop condition when PHP 8.0+ is required
```
and most of them share the same problem - using undefined constants (like `T_ENUM` in PHP 7.4) even in data provider fails the PHPUnit.

File: tests/Tokenizer/Analyzer/Analysis/TypeAnalysisTest.php (hunk 6 above)
Comment (kubawerlos) [reply tests/Tokenizer/Analyzer/Analysis/TypeAnalysisTest.php:175]: This was not moved anywhere - this PR adds the check and fixes all non-fixer tests classes.

File: tests/AutoReview/ProjectCodeTest.php (hunk 5 above)
Comment (kubawerlos) [reply tests/AutoReview/ProjectCodeTest.php:826]: For this single cache variable only? If we want to do it let's do it at least in all `AutoReview` tests (in separate PR and with measuring if it has an impact on performance).

File: tests/Tokenizer/Transformer/TypeIntersectionTransformerTest.php (hunk 3 above)
Comment (keradus) [reply tests/Tokenizer/Transformer/TypeIntersectionTransformerTest.php:66]: most/all of them can be replaced with checking PHP version.
OK to have it in another iteration.

File: tests/AutoReview/ProjectCodeTest.php (hunk 5 above)
Comment (keradus) [reply tests/AutoReview/ProjectCodeTest.php:826]: I wonder when the garbage collector clear it, if any. back in time PHPUnit was _not_ unregistering instances of test classes, eg `ProjectCodeTest`, and `static` variable remained defined forever, keeping the place in memory.

File: tests/Tokenizer/Analyzer/Analysis/TypeAnalysisTest.php (hunk 6 above)
Comment (keradus) [reply tests/Tokenizer/Analyzer/Analysis/TypeAnalysisTest.php:175]: I thought you started moving those things in #7581 ?

File: tests/AutoReview/ProjectCodeTest.php (hunk 5 above)
Comment (keradus) [reply tests/AutoReview/ProjectCodeTest.php:826]: @kubawerlos ,

`static $foo` within a method works exactly same way as within a function. It creates global object. PHP will have no clue when to free memory of this global object, because it doesn't know if it will be used again later.

And it behave same way as in function, meaning that it's state is shared within all instances - including future instances. Thus even if TestCase class instance would get `unset()`, PHP will still keep static(==global) variable in memory, in case any other instance would be created and method called, eg https://3v4l.org/3iRb0

moving this to static property of class allows us to instruct PHP to mark this property as free-able by GC, by putting `= null` instruction in `tearDownAfterClass`.

---

and here, you create , in that global variable, over 1_000 tokens collections

**Review by keradus — CHANGES_REQUESTED** [review]
blocking to prevent merge before conclusion. I'm not convinced this is good path to replace #7575 , I would love to have tests cleaned up first and see how many, if any, test files contains PHP_* outside of provider.

File: tests/Tokenizer/Analyzer/Analysis/TypeAnalysisTest.php [hunk 8]
```diff
@@ -123,75 +126,120 @@ public static function provideIsNullableCases(): iterable
...
 
-        if (\PHP_VERSION_ID >= 8_02_00) {
-            yield [true, '(\foo\bar&\foo\baz)|null'];
+        yield [false, '\foo\bar&Null2'];
+    }
+
+    /**
+     * @dataProvider provideIsNullable82Cases
+     *
+     * @requires PHP 8.2
```
Comment (keradus) [inline tests/Tokenizer/Analyzer/Analysis/TypeAnalysisTest.php:220]: this could also got moved to 7581 🤷🏻

Comment (kubawerlos) [reply tests/Tokenizer/Analyzer/Analysis/TypeAnalysisTest.php:220]: Do you mean twice? 😉 

https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7581/files#diff-3c8f800dccefd2fcba25f19e94fe2aa6e02b4a86c55045e1cb407b46cd7b773eR220

Comment (keradus) [reply tests/Tokenizer/Analyzer/Analysis/TypeAnalysisTest.php:220]: i assumed logic was moved, not copied 🤷🏻
